### PR TITLE
CI: Travis -> GitHub Actions; Create Release Binaries and Container Images

### PR DESCRIPTION
**Pull request by sheurich**
[Checks](https://github.com/fastly/pebble/actions/workflows/checks.yml)
[Tests](https://github.com/fastly/pebble/actions/workflows/tests.yml)

As a follow-on to https://github.com/letsencrypt/pebble/pull/439 this change request implements the first step of building a new CI system for Pebble. As this is a testbed environment for Boulder, some of these changes may prove to be useful in the latter context.

Per https://github.com/letsencrypt/pebble/issues/434#issuecomment-1958432503, GitHub Actions is a reasonable choice for a new CI implementation to supersede Travis. Much, but not all, of the existing functionality is present in this initial set of changes.

Items currently implemented include:
- Binary production of `pebble` and `challtestsrv` for platforms:
  - AMD64
    - darwin
    - linux
    - windows
  - ARM64
    - darwin
    - linux
    - windows
- Multiplatform docker images published on the GitHub Container Registry for platforms:
  - linux/amd64
  - linux/arm64
  - windows/amd64
- Golang linting.
- Golang coverage.
- Local test scripts for Go and Docker:
  - `./build.sh`
  - `./docker.sh`
  - `./test.sh`

This PR also adds a `-version` flag to Pebble, which is set during release build to the Git commit ref.

This change builds on earlier work done in https://github.com/letsencrypt/pebble/pull/379 and https://github.com/letsencrypt/pebble/pull/442.

### CODE REVIEW

**Comment by sheurich** [comment]
Example workflow run can be seen in:
[CI](https://github.com/fastly/pebble/actions/workflows/ci.yml)

**Comment by ldez** [comment]
I think we can merge our 2 PRs #442

**Comment by sheurich** [comment]
> I think we can merge our 2 PRs #442

@ldez Yes let's do that. I can take a look tomorrow.

**Comment by ldez** [comment]
I mean I think my PR can be merged and your PR will just need to be "cleaned" to remove the duplication with mine.

**Comment by sheurich** [comment]
@ldez glad you found https://github.com/letsencrypt/pebble/issues/440! I spent some time troubleshooting at the CI level but hadn't checked Pebble itself yet.

File: .github/workflows/ci.yml
Comment (ldez) [inline .github/workflows/ci.yml:116]: I think you can drop this part because it's already handled by the other workflow.

File: .github/workflows/build.yml
Comment (ldez) [inline .github/workflows/build.yml:8]:
```suggestion
  push:
    tags:
      - v*
```

this workflow uses a secret and publish some elements. 

and I recommend to restore my previous file `go-cross.yml`

File: .github/workflows/checks.yml
Comment (ldez) [inline .github/workflows/checks.yml:36]: I'm the main maintainer of golangci-lint, and based on this repo activity I recommend not using the latest.
The action can have some issues with the cache.

File: .github/workflows/checks.yml
Comment (ldez) [inline .github/workflows/checks.yml:21]: It is better for the maintenance to have all the configurable elements in one place.

File: .github/workflows/build.yml
Comment (sheurich) [reply .github/workflows/build.yml:8]: Thanks for the quick feedback! What should be different in the PR build process versus release? Is the OS matrix in go-cross intended to verify build success for common platforms, whereas the release output can come from a single build platform?

File: .github/workflows/checks.yml
Comment (sheurich) [reply .github/workflows/checks.yml:21]: Why use oldstable for these checks rather than stable?

File: .github/workflows/build.yml
Comment (ldez) [reply .github/workflows/build.yml:8]: The release will publish artifacts (archive, GitHub release, Docker images, etc.), a PR will not do that.
A PR doesn't need to build Docker images or create archives.
It's better to not create one workflow with the conditions but to have a dedicated workflow by topic.

In general, nobody should push on a repo without creating a PR before, so the workflow of the PR will check tests, linting, and binary builds.
So when you merge a PR, tests, and linting are useless because they have been already checked.

This is the same thing for builds: the build for a release is not just binary but a set of builds specific to a release (archive, GitHub release, Docker images, etc.).

File: .github/workflows/checks.yml
Comment (ldez) [reply .github/workflows/checks.yml:21]: Because the previous CI was doing that.
You can use `stable`, it's not a problem for me, but I still recommend centralizing the variables, it helps for the maintenance and to diagnose CI problems.

File: .github/workflows/checks.yml
Comment (ldez) [reply .github/workflows/checks.yml:36]: I know but I also know the problem behind the action.
This is why the previous workflow doesn't it.
If you to use it, you should disable the cache:
```yml
    skip-cache: true
    skip-build-cache: true
```

**Comment by sheurich** [comment]
@ldez I think all of the changes have been made and would appreciate another review. The PR description hasn't yet been updated to match the actual changes.

File: .github/workflows/build.yml
Comment (ldez) [inline .github/workflows/build.yml:1]: maybe rename to `release.yml`

**Comment by sheurich** [comment]
This will be ready for review soon, pending a bug fix.

**Comment by sheurich** [comment]
I think this should be ready for review @mcpherrinm

**Comment by mcpherrinm** [comment]
It doesn’t have to be in this PR, but we are generally standardizing on either distroless or debian-slim as the base for docker images instead of alpine.  We can make that change later, or it would be helpful if you want to make it here while we’re changing the docker files.

I’ll review more fully later today

**Comment by sheurich** [comment]
> It doesn’t have to be in this PR, but we are generally standardizing on either distroless or debian-slim as the base for docker images instead of alpine.

I had assumed that Alpine was chosen over Debian as a derivative of the strategy outlined in the Pebble README:
> In places where the ACME specification allows customization/CA choice Pebble aims to make choices different from Boulder.

It's no problem to change to Debian for dev and Distroless for release.

**Comment by sheurich** [comment]
Since `cgo` is not needed by pebble or pebble-challtestsrv, the release image can be `FROM scratch`. This WFM but are there other requirements which need glibc or other items from distroless?

**Comment by mcpherrinm** [comment]
Scratch is fine by me

**Comment by sheurich** [comment]
@mcpherrinm the `scratch` change is implemented and a few other bugs have been squashed.

File: .github/workflows/release.yml
Comment (aarongable) [inline .github/workflows/release.yml:10]: I don't think this one needs access to PRs, since it's only running on tag pushes

**Review by aarongable — COMMENTED** [review]
The four shell scripts (docker, lint, mods, and test) seem like they'll be kinda difficult to keep in sync with the commands in the corresponding github actions. Can we either have the github actions call these scripts directly, or leave them out, or clearly document in both places how best to keep them in sync?

**Comment by sheurich** [comment]
> The four shell scripts (docker, lint, mods, and test) seem like they'll be kinda difficult to keep in sync with the commands in the corresponding github actions.

I agree.

> Can we either have the github actions call these scripts directly, or leave them out, or clearly document in both places how best to keep them in sync?

I found them useful and like the idea of being able to run the same checks/tests as CI locally. I will update the workflows to call the scripts.

**Review by aarongable — APPROVED** [review]

**Review by mcpherrinm — APPROVED** [review]

**Comment by mcpherrinm** [comment]
Merged and tagged as v2.5.0

**Pull request merged**
